api/stats.js

```diff
@@ -1,31 +1,73 @@
 import { config } from './../config.js';
 import { SnowflakeUtil, MessageEmbed } from "discord.js";
 
+const removeParentheses = word => word.replace(/\(|\)/g, '');
 export class Stats {
     constructor(discordClient) {
         this.client = discordClient;
         this.stats = {};
     }
 
-    initializeStats(prev_message) {
-        /*
-        This is a placeholder.
-        We have some limitations on how many messages we can load per request
-        we can have daily pulls, use the latest message as the last source of truth
-        and update with new stats
-        let stats = {
-            username1: {
-                messages: 5, 
-                words: 20, 
-                letters: 150
+    /*
+        Returns a snowflake which represents the time x days in the past
+        https://discord.com/developers/docs/reference#snowflakes
+    */
+    _generatDaySnowflake(days) {
+        let d = new Date();
+        d.setHours(0, 0, 0, 0);
+        d.setDate(d.getDate() + days);
+        return SnowflakeUtil.generate(d);
+    }
+
+    async _getStatsChannel() {
+        let dominationGuild = this.client.guilds.resolve(config.GUILD_ID);
+        return dominationGuild.channels.cache.find(
+            (channel) => channel.name === config.STATS_CHANNEL
+        );
+    }
+
+    async _computeWeeklyStats() {
+        let sevenDaysAgo = this._generatDaySnowflake(-7);
+        let statsChannel = await this._getStatsChannel();
+
+        let messages = await statsChannel.messages.fetch({ limit: 100, after: sevenDaysAgo });
+        let collectedDays = new Set();
+        let stats = {};
+
+        for (let message of messages.values()) {
+            let isBot = message.author.username === 'elprimobot';
+            let isDaily = message.content.indexOf('Daily') > -1;
+            let day = new Date(message.createdTimestamp);
+            let key = `${day.getDay()}/${day.getMonth()}/${day.getFullYear()}`;
+            // we only take the latest stat posted for the bot that day
+            if (!isBot || !isDaily || collectedDays.has(key)) {
+                continue;
+            }
+            collectedDays.add(key);
+
+            let fields = message.embeds[0].fields;
+            let firstColumn = fields[0].value.split('\n');
+            let secondColumn = fields[2].value.split('\n');
+            console.log(firstColumn, secondColumn);
+            let i = 0;
+            while (i < firstColumn.length) {
+                let [username, posts] = firstColumn[i].split(' ');
+                let [words, letters] = secondColumn[i].split(' ');
+                stats[username] = stats[username] || { posts: 0, words: 0, letters: 0 };
```

Comment:
It would be better to update the stats object treating it as an immutable object, eg: stats = {...stats, [username]: {...stats[username], posts: stats[username]+= Number(removeParentheses)}}

Reply:
I'm not convinced there are a lot of benefits here, we're basically creating this object in this function - same object, I can def follow the immutable way but I think in this case, i think it'll just make it harder to read while not adding real benefits, I think it'd be nice to freeze the object once we return to get some kind of benefits with full immutability.